lib/validators/one_favorite_per_owner_validator.rb

```diff
@@ -0,0 +1,16 @@
+module Validators
+  # Validator for Collection model
+  # Validate that only one favorite collection is created per owner for each project
+  class OneFavoritePerOwnerValidator < ActiveModel::Validator
+    def validate(new_record)
+      return unless new_record.owner
+
+      records = new_record.owner.collections
+      return unless records && new_record.favorite
+
+      records.each do |record|
+        new_record.errors[:favorite] = 'Only one favorite collection can be created per owner for each project' if record.favorite
```

Comment:
I think you can optimize this looping to a db call above where your find all the owner collections. instead of finding them all and iterating over them to check if any are favs we can just ask the db if there are any and mark as error if there are. Instead of this loop where you do the record fetches try something like this using https://apidock.com/rails/v4.2.9/ActiveRecord/FinderMethods/exists%3F
```
has_existing_favorite = new_record.owner.collections.where(favorite: true).exists?
# then you can mark with error
if has_existing_favorite 
  new_record.errors[:favorite] = 'Only one favorite collection can be created per owner for each project'
end
```

Reply:
Wow yes, this is much cleaner.